### PR TITLE
Factor out object read/write from BucketAlertStore into ProtobufStore.

### PR DESCRIPTION
**Pull request by stevesg**
**What this PR does**:
Factor out object read/write from BucketAlertStore into ProtobufStore. The BucketAlertStore is 90% of what I need for persisting the notifications/silences state, so it seems wasteful to copy/paste it.

### CODE REVIEW

File: pkg/alertmanager/alertstore/bucketclient/protobuf_store.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	"github.com/gogo/protobuf/proto"
+	"github.com/thanos-io/thanos/pkg/objstore"
+	"github.com/thanos-io/thanos/pkg/runutil"
+
+	"github.com/cortexproject/cortex/pkg/storage/bucket"
+)
+
+// ProtobufStore is used to read/write protobuf messages to object storage backend. It is implemented
+// using the Thanos objstore.Bucket interface
+type ProtobufStore struct {
```
Comment (stevesg) [inline pkg/alertmanager/alertstore/bucketclient/protobuf_store.go:18]: Naming suggestions welcome.

File: pkg/alertmanager/alertstore/bucketclient/protobuf_store.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
+package bucketclient
```
Comment (stevesg) [inline pkg/alertmanager/alertstore/bucketclient/protobuf_store.go:1]: I assume there's a better place to put this file, perhaps somewhere in `pkg/storage` or `pkg/util`, just tell me where and I'll move it.

File: pkg/alertmanager/alertstore/bucketclient/protobuf_store.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	err := s.bucket.Iter(ctx, "", func(key string) error {
+		userIDs = append(userIDs, key)
+		return nil
+	})
+
+	return userIDs, err
+}
+
+// Set writes the content for an entry in the store for a particular user.
+func (s *ProtobufStore) Set(ctx context.Context, userID string, msg proto.Message) error {
```
Comment (pracucci) [inline pkg/alertmanager/alertstore/bucketclient/protobuf_store.go:45]: What do you think about calling it `Upload()` to keep it specular to the underlying `objstore.Bucket` function name?

**Review by pracucci — COMMENTED** [review]
Thanks for working on this refactoring.

I left a couple of comments but then I stopped reviewing it cause I had a doubt. The structure of the alertmanager config in the bucket is `alerts/<user id>` where `alerts/<user id>` is the object key storing the alertmanager config for `<user id>`.

We're not used to have this pattern for the storage, while we typically have `alerts/<user id>/<object name>` structure. When we designed the state persistence, we agreed on using this latter structure, but I've the feeling it will clash with this refactoring, because `ProtobufStore` is designed around the alertmanager structure which is a special case.

I would like to better understand how you plan to use `ProtobufStore` for the state persistence before continuing with the review. Thanks!

**Comment by stevesg** [comment]
After talking to @pracucci, I like the idea of adding the interfaces onto the existing `AlertStore` instead. This will save a some of plumbing (passing in a new store or the configuration required for the store), and makes sense if we want to store them in the form `alerts/<user id>/<object name>` alongside the existing configuration state.

I've move this to draft for now - no need to review further.

**Comment by stevesg** [comment]
Closing this in preference of https://github.com/cortexproject/cortex/pull/4020

**Pull request closed without merging**